### ISSUE TITLE
marker.setParentFrame() and marker.setFrameName() behave inconsistently

### ISSUE THREAD

**Issue by aymanhab** [opensim-org/opensim-core#1889]
newMarker.setParentFrame(frameObject); works when using an actual Frame object  but newMarker.setFrameName("parentName") fails to connect in initSystem complaining that initSystem must be called first! We need to either make them consistent or document why they are different and improve the error message so it's not circular.

**Comment by aseth1** [comment]
This is the code for `setFrameName()`
```cpp
void Marker::setFrameName(const string& name)
{
    const auto& refFrame = getModel().getComponent<PhysicalFrame>(name);
    setParentFrame(refFrame);
}
```

**Comment by aseth1** [comment]
Calling `getModel().getComponent<PhysicalFrame>()` would be an issue with an unconnected marker where `getModel()` would throw (likely asking if you forgot to call `initSystem()` first) but would not throw in `initSystem()` AFAICT.

**Comment by aseth1** [comment]
I also think the redundancy is not helpful, so I am happy to get rid of `setFrameName()` which was a carry over from `setBodyName()`.

**Comment by aymanhab** [comment]
On one side I'd prefer to have it removed since its behavior is different (requires some connection which is not documented anywhere AFAIK) even though it looks very similar from an API user perspective to setFrame, which is unexpected. On the other hand it seems more natural to create an object using primitive data types (e.g. strings for names) rather than having to get a reference to a valid Frame to construct/add a Marker.

What would be your preference as an API user, @chrisdembia @tkuchida ?

**Comment by aseth1** [comment]
We can update `Marker::setFrameName()` to only set the underlying Socket's connectee_name (like what happens by deserializing) and establish the connection at `finalizeConnections()`. This allows the model creator to defer the validation of connections, but it also means you don't have a "wired-up" connection. It seems the creator of Marker wanted the `setFrameName()` to result in a live connection and that's why it gets the `Frame` and calls `setParentFrame()`  with it.

While setting the Component directly does not allow you to defer creation of Components and their validation (you either have a valid Component or not) it is logically reasonable expect Components you want to connect to be available. That means potentially reordering the order of operations in user-code (although I cannot think of any of our examples that require the change) to create the dependencies before connecting them. If one insists on predefining the connections prior to the creation of the Components that satisfy them (the connectee) s/he can always use `updSocket<C>("socket_name").setConnecteeName("comp_name")`. 

Long term, I am inclined to remove all `set<Connectee>Name()` methods from the API moving towards a minimally complete interface. A first step would be to deprecate all `set<Connectee>Name()` and make sure they only  `updSocket<C>("socket_name").setConnecteeName("C_name")`.  One can always locate a Frame (or any Component) by (path) name OR create a `Frame` on the fly and call the convenience `setFrame()`. 

IMHO it is a cleaner more robust API to deal with Components than strings. @chrisdembia @tkuchida @jenhicks @carmichaelong please let me know your thoughts as well.

**Comment by aymanhab** [comment]
Thanks for the investigation @aseth1 I'd lean to set the name property immediately and to delay connecting till later, since this allows setting multiple, or possibly all connections at once (and follows the same paradigm as post deserialization) as opposed to users guessing what part of the model has been connected and trying to get a reference (which may fail) to pass down.. While cleaner, using references shifts the mental overhead of figuring out what's connected to the user unnecessarily. If we had clear "stages" then at least a user could query the model to see where it's at. Is this possible with the current API? Or maybe I'm just missing a part of the picture here.

**Comment by tkuchida** [comment]
I agree with @aymanhab: `setFrameName(const string& name)` should do no more than set the appropriate `string` property. The current implementation @aseth1 quoted above (second comment) looks like it should be an overload of `setParentFrame()` that takes a string rather than a reference to a `PhysicalFrame`. I disagree with removing this and similar methods, as I think `setParentFrame()` will be more intuitive to users than `updSocket<PhysicalFrame>("parent_frame").setConnecteeName(name)`.

**Comment by chrisdembia** [comment]
I agree with @tkuchida

**Comment by aseth1** [comment]
> A first step would be to deprecate all set<Connectee>Name() and make sure they only updSocket<C>("socket_name").setConnecteeName("C_name").

OK, I won't deprecate, for now. I'll make sure the connection property setting by name is consistent throughout. I will make that change for 4.0 so at least `setFrameName()` and others can be used consistently. I will also rename to `setParentFrameName()` to make it consistent with `setParentFrame()`.
 
I disagree with the overload to use a string to set a live connection (as is attempted) in the current implementation for minimal completeness and, IMHO, it is better to distinguish a live connection using a reference (`setParentFrame()`) VS a convenience for setting the connection property  (and delayed connection) using  `setParentFrameName()`.

**Comment by aseth1** [comment]
closed by #1901
